### DISCUSSION TITLE
** Feature in splitCustom.c

### DISCUSSION THREAD

**Discussion by yuw444** [kogalur/randomForestSRC#197] [Q&A]
Hi Dr. Ishwaran,

I have some questions about customizing the splitting rule. 

1.  How to pass variables(not necessarily x variables but same length with the sample size) to `**Feature` in `splitCustom.c`? Later, use those features for building the splitting statistics. 
2. In this case, are `**Feature` changing dynamically in each node? 

Thanks

**Comment by kogalur** [comment]
In splitCustom.c we describe the feature matrix:

  feature - matrix of user specified features that can be sent into the split rule and acted on as desired.  The matrix is of dimension
  [featureCount] x [n].  Features are neither y-variables or x-variables. However, for expediency, they are specified by the user
  as y-variables, but are tagged as having zero weight via the y-variable weight vector.  Thus, they are never used in the the
  pre-defined split rules, and have no predicted value. The pointer will be NULL when features are absent.

  featureCount - count of features in the above matrix, specifically the number of rows.  The count will be zero when features are absent.

So adjust the data frame to include the feature matrix, and specify your features as y-variables, with zero weight. In a node, those features in the node, can be used in the split rule. They change in each node according to the membership. If record i is in the node, then the set of features in record i are also in that node and can be used by the split rule.

In the same way response[i] represents the response for record [i], feature[j][i] is the double value corresponding to the [j]th feature for record [i]

**Reply by yuw444** [reply]
Hi Kogular, Thanks for your prompt response. I have gone over the description several times, still feel confused.

In particular, how can I pass the feature matrix in R interface? In `rfsrc( )`, `y` is inside `formula` argument, so, I am not sure how to assign a matrix to `y`. Or, I am in a completely wrong direction. Is there any other argument I can use in `rfsrc()`? 

Additionally, zero weighted y-variables, do you mean to adjust the weight via `yvar.wt` in `rfsrc()`? But, it said "NOT YET IMPLEMENTED" in R help file. 

Thanks

**Reply by yuw444** [reply]
Do you mean something like this? https://luminwin.github.io/randomForestSRC/articles/mvsplit.html, should I set `yvar.wt = c(1,rep(0,20))` if I like only `C14.0` as my true y variable and other as the feature matrix?

**Reply by kogalur** [reply]
Sorry for the delay in this response.  Here is a canonical example that you can use:

Take the data set mtcars as our example.  This data set can use the existing custom split rule is splitCustom.c. It can be accessed via the following:

mtcars.reg = rfsrc(mpg ~., data = mtcars, splitrule="custom1")

Now, say you want to send in features.  Let's just send in "cyl" as a feature instead of as an x-variable.  We would do this as follows:

mtcars.mreg = rfsrc(Multivar(mpg, cyl) ~ ., data = mtcars, splitrule = "custom1", yvar.wt = c(1,0))

In splitCustom.c you would see that the function:

getCustomSplitStatisticMultivariateRegression ()

would make available the node specific data related to "cyl" via the last incoming parameters

double **feature
unsigned in featureCount

In this case, featureCount = 1 because we have only one feature, namely "cyl".  The parameter double **feature is a [1] x[n] array where n is the population in the node and is given by the first parameter in that function:

unsigned int n

So feature[1][...] would consist of 4's, 6's, and 8's and you would use these values as you wished.  You could ignore them entirely.  If you ignored them, they would not be used in the split rule. This would be the equivalent of growing a forest without "cyl".  So you should find that the following two calls are equivalent:

## Omit "cyl" from the univariate regression data set.
mtcars.reg = rfsrc(mpg  ~ disp + hp + drat + wt + qsec + vs + am + gear + carb, data = mtcars, seed = -1, mtry = Inf, nsplit = 0)

## Send in "cyl" as a feature via multivariate regression, but ignore it.
mtcars.mreg = rfsrc(Multivar(mpg, cyl) ~., data = mtcars, splitrule="custom1", yvar.wt=c(1,0), seed = -1, mtry = Inf, nsplit = 0)

**Reply by yuw444** [reply]
Wow, very in-depth explanation. I'm very clear now. Thanks so much.  Hope this will help others too.

**Reply by yuw444** [reply]
Additionally, can I define a constant vector that doesn't change as the tree grow and can be used in the splitting rule?

**Comment by kogalur** [comment]
Yes, this would be perfectly fine.  It would be a way to grow different models using different constant vectors while using the same complied custom split rule.

**Reply by yuw444** [reply]
Sorry, I didn't make my question clear. I like to keep a user-defined constant in the splitting rule, it could be a scalar or vector. WLOG, alpha = c(1,2,3,4) in the splitting rule, is there any way I can make this happen? Thanks so much for your help!

**Comment by kogalur** [comment]
Let's say you have a vector of user defined constants alpha = c(alpha1, alpha2, alpha2) that you want to make available to your custom split rule.  Again, let's take mtcars, and splitrule = "custom1" in the package as our starting point.  First, you would create an auxiliary data frame as follows:

aux.response = as.data.frame(cbind(rep(alpha1, dim(mtcars)[1]), rep(alpha2, dim(mtcars)[1]), rep(alpha3, dim(mtcars)[1])))
colnames(aux.response) = c("alpha1", "alpha2", "alpha3")

Then create an augmented data set as follows:

mtcars.aux = cbind(aux.response, mtcars)

Then a statement like 

mtcars.aux.mreg = rfsrc(Multivar(mpg, alpha1, alpha2, alpha3) ~ ., data = mtcars.aux, splitrule = "custom1", yvar.wt = c(1, 0, 0, 0))

would make available the feature as a  [3] x[n] array in each node in the split rule.  It's a bit redundant, because each row is a constant containing alpha1, alpha2, and alpha3, but you could use these constants as you wish.  This is what I meant by saying that you would only have to compile once, and introduce the constants via an augmented data frame.

**Reply by yuw444** [reply]
I see your point now.  

To be honest, I am doing my dissertation which is highly related to this modification. Right now, it's under testing, not sure how it will go. But I have faith in it. 

I'm wondering if you heard [rfSLAM](https://github.com/mattrosen/rfSLAM), which is built upon your R package. I'm trying to do something similar to what they did. I have almost done with `splitCustom.c`, but need some additional user-defined variables(vector) available for the splitrule.  So, it may request some customization on your original C code. Here, may I make bold to ask your help(collaboration)? If you are interested, we can discuss more details.

Thanks so much for all of your patient help.

**Reply by yuw444** [reply]
> Let's say you have a vector of user defined constants alpha = c(alpha1, alpha2, alpha2) that you want to make available to your custom split rule. Again, let's take mtcars, and splitrule = "custom1" in the package as our starting point. First, you would create an auxiliary data frame as follows:
> 
> aux.response = as.data.frame(cbind(rep(alpha1, dim(mtcars)[1]), rep(alpha2, dim(mtcars)[1]), rep(alpha3, dim(mtcars)[1]))) colnames(aux.response) = c("alpha1", "alpha2", "alpha3")
> 
> Then create an augmented data set as follows:
> 
> mtcars.aux = cbind(aux.response, mtcars)
> 
> Then a statement like
> 
> mtcars.aux.mreg = rfsrc(Multivar(mpg, alpha1, alpha2, alpha3) ~ ., data = mtcars.aux, splitrule = "custom1", yvar.wt = c(1, 0, 0, 0))
> 
> would make available the feature as a [3] x[n] array in each node in the split rule. It's a bit redundant, because each row is a constant containing alpha1, alpha2, and alpha3, but you could use these constants as you wish. This is what I meant by saying that you would only have to compile once, and introduce the constants via an augmented data frame.

Can I have a similar implementation with Surv() in survival setting?

**Reply by kogalur** [reply]
Yes, the technique works for all families. There are sample harnesses for all families in the custom split rule code.

**Reply by yuw444** [reply]
I kinda get it. However, I couldn't set up "multiple y's" in the formula `Surv( )~.`.

Also, I checked the source code `randomForestSRC:::parseFormula`, it seems only allow 2 variables or 4 variables in `yvar.names` for the survival family. Is there any way to get around it?  If possible, could you give a toy example? Thanks so much for your time.

**Reply by yuw444** [reply]
I saw there are `userFeature` in both `char customSurvivalSplit` and `char customCompetingRiskSplit`, just couldn't figure out how to set them up correctly without hurting the later `predict()`.

**Reply by kogalur** [reply]
Our apologies, the above answer to your query was incorrect.  We have not yet implemented the ability to send in features with survival and competing risks. It's on our list of campaigns.  Unfortunately, we can't give you an estimate of when that will be completed. , but we will let you know.

**Reply by yuw444** [reply]
No problem. I will keep my eyes on it. Thanks.

**Comment by camposacrs** [comment]
Hi Drs. Ishwaran and Kogalur,

As I could follow from this discussion, it is not possible to send features in custom splitting rules for survival and competing risks families.  

I'm writing here beacause for the specific case of random survival forests, I need a custom splitting rule that uses a matrix of features. To calculate  it I need to access some individual variables, such as sex and age.  However, using only the inputs passed as arguments in the **CustomSplitStatisticSurvival** it is not possible to do it without the feature matrix.  

Given a vector of **membership**, **time** or **event** (that are the function inputs associated to a **node**), is it possible to assess the other variables associated to each of these individuals (like age, sex, etc), so that I can implement my feature matrix inside the CustomSplitStatisticSurvival function?

Thank you in advance for the answer.

**Comment by kogalur** [comment]
Custom features are used and passed in univariate classification and regression families, and multivariate families only.  Unfortunately, to allow this in survival and competing risk requires decent modification of the way we pass these custom features, both on the R-side and how we access the features on the C-side.  At this time, we do not have a time-line of when this particular campaign will be completed.  We apologize for not being more informative, at this point, but it will be possible to do it in the future.

**Comment by camposacrs** [comment]
Hi Mr kogalur,
Thank you for your answer.

My interest is to use an appropriate spitting rule to build a forest for a particular survival measure. For that, I will need some features that depends on demographic covariates, which I would like to pass as a feature in the custom splitting rule function.  Since my interest at this moment is only the forest building, is it a possible to consider an adpatation using the Multivariate specification with  a formula that pass (i) the time-to-event as a response and (ii) the event indicator and the constant matrix that I need as a feature (using the argument yvar.wt that you meention).  Then, inside the custom function, I would compute the quantities of interest for the survival context (for example, to calculate a log-rank type statistics). 

Is that type of  adaptation  OK, or is there anything conceptually wrong in that practice (considering the structure of  the implemented functions of the randomForestSRC package)?
